File: app/src/main/java/com/sjn/stamp/media/CustomController.java
package com.sjn.stamp.media;

import android.content.Context;

import com.sjn.stamp.constant.ShuffleState;
import com.sjn.stamp.controller.UserSettingController;
import com.sjn.stamp.utils.LogHelper;
import com.sjn.stamp.constant.RepeatState;

import java.util.ArrayList;
import java.util.List;

public class CustomController {
    private static final String TAG = LogHelper.makeLogTag(CustomController.class);

    private static CustomController sInstance;

    private List<ShuffleStateListener> mShuffleStateListenerSet = new ArrayList<>();

    private List<RepeatStateListener> mRepeatStateListenerSet = new ArrayList<>();

    public ShuffleState getShuffleState() {
        return mShuffleState;
    }

    public RepeatState getRepeatState() {
        return mRepeatState;
    }

    private ShuffleState mShuffleState;

    private RepeatState mRepeatState;

    public interface RepeatStateListener {
        void onRepeatStateChanged(RepeatState state);
    }

    public interface ShuffleStateListener {
        void onShuffleStateChanged(ShuffleState state);
    }

    public void addShuffleStateListenerSet(ShuffleStateListener listener) {
        mShuffleStateListenerSet.add(listener);
    }

    public void removeShuffleStateListenerSet(ShuffleStateListener listener) {
        mShuffleStateListenerSet.add(listener);
    }

    public void addRepeatStateListenerSet(RepeatStateListener listener) {
        mRepeatStateListenerSet.add(listener);
    }

    public void removeRepeatStateListenerSet(RepeatStateListener listener) {
        mRepeatStateListenerSet.add(listener);
    }

    public void setRepeatState(Context context, RepeatState state) {
        mRepeatState = state;
        for (RepeatStateListener repeatStateListener : mRepeatStateListenerSet) {
            repeatStateListener.onRepeatStateChanged(state);
        }
        UserSettingController userSettingController = new UserSettingController();
        userSettingController.setRepeatState(state);
    }

    public void setShuffleState(Context context, ShuffleState state) {
        mShuffleState = state;
        for (ShuffleStateListener shuffleStateListener : mShuffleStateListenerSet) {
            shuffleStateListener.onShuffleStateChanged(state);
        }
        UserSettingController userSettingController = new UserSettingController();
        userSettingController.setShuffleState(state);
    }

    public void toggleRepeatState(Context context) {
        setRepeatState(context, mRepeatState.toggle());
    }

    public void toggleShuffleState(Context context) {
        setShuffleState(context, mShuffleState.toggle());
    }

    public static CustomController getInstance() {
        if (sInstance == null) {
            sInstance = new CustomController();
        }
        return sInstance;
    }

    private CustomController() {
        mShuffleState = ShuffleState.Companion.getDefault();
        mRepeatState = RepeatState.Companion.getDefault();
    }

}

File: app/src/main/java/com/sjn/stamp/media/CustomController.kt
package com.sjn.stamp.media

import com.sjn.stamp.controller.UserSettingController
import com.sjn.stamp.model.constant.RepeatState
import com.sjn.stamp.model.constant.ShuffleState
import java.util.*

object CustomController {

    private val shuffleStateListenerSet = ArrayList<ShuffleStateListener>()
    private val repeatStateListenerSet = ArrayList<RepeatStateListener>()

    var shuffleState: ShuffleState = UserSettingController().shuffleState
        private set(state) {
            field = state
            for (shuffleStateListener in shuffleStateListenerSet) {
                shuffleStateListener.onShuffleStateChanged(state)
            }
            val userSettingController = UserSettingController()
            userSettingController.shuffleState = state
        }

    var repeatState: RepeatState = UserSettingController().repeatState
        private set(state) {
            field = state
            for (repeatStateListener in repeatStateListenerSet) {
                repeatStateListener.onRepeatStateChanged(state)
            }
            UserSettingController().repeatState = state
        }

    interface RepeatStateListener {
        fun onRepeatStateChanged(state: RepeatState)
    }

    interface ShuffleStateListener {
        fun onShuffleStateChanged(state: ShuffleState)
    }

    fun addShuffleStateListenerSet(listener: ShuffleStateListener) =
            shuffleStateListenerSet.add(listener)

    fun removeShuffleStateListenerSet(listener: ShuffleStateListener) =
            shuffleStateListenerSet.add(listener)

    fun addRepeatStateListenerSet(listener: RepeatStateListener) =
            repeatStateListenerSet.add(listener)

    fun removeRepeatStateListenerSet(listener: RepeatStateListener) =
            repeatStateListenerSet.add(listener)

    fun toggleRepeatState() {
        repeatState = repeatState.toggle()
    }

    fun toggleShuffleState() {
        shuffleState = shuffleState.toggle()
    }

}